### PR TITLE
Initial support for BFRuntime Json Generation for DPDK

### PR DESCRIPTION
**Pull request by amreshk**
In order to develop Tofino-specific P4 data plane programs, a key component is generating information compatible with BFRuntime. More details are listed in this repo https://github.com/barefootnetworks/Open-Tofino.

This PR adds support to generate BFRuntime json files for DPDK backend. Output json is in a table driven interface format supported by BFRuntime. 

Support adds the following
- Common code in `control_plane` folder `bfruntime.h/cpp` to generate json file for default PSA arch. This code can be used and extended for any backend
- Enabling DPDK to generate json and adding DPDK extensions for ActionSelector extern (dpdk specific proto file). This showcases how backend specific externs can be added by extending the P4RuntimeArchHandler (`bfruntime_arch_handler`) and generating the extern specific tables in json `bfruntime_ext.h/cpp` 
- Some refactoring of `P4RuntimeArchStandard` code to allow for extending capabilities

Misc fixes
- Fixed a PSA bug on ActionSelector and ActionProfile which did not output the respective `implementation_id` and `table_id` in p4info. This was due to incorrectly checking for `implementation` string instead of `psa_implementation`.
- Above fix exposed an unsupported issue in `psa-action-selector2.p4` which had multiple action selectors in the `psa_implementation` string. Currenttly a list of expressions is not supported. This is marked as xfail to be fixed outside this PR.

### CODE REVIEW

**Review by hanw — COMMENTED** [review]
Ping @antoninbas, please take a look at the refactoring of p4RuntimeArchStandard.cpp/h, thanks.

File: backends/dpdk/control-plane/bfruntime_arch_handler.h [hunk 1]
```diff
@@ -0,0 +1,272 @@
+#ifndef DPDK_CONTROL_PLANE_BFRUNTIME_ARCH_HANDLER_H_
```
Comment (mbudiu-vmw) [inline backends/dpdk/control-plane/bfruntime_arch_handler.h:1]: no copyright headers?

File: backends/dpdk/control-plane/bfruntime_ext.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
+#include "bfruntime_ext.h"
```
Comment (mbudiu-vmw) [inline backends/dpdk/control-plane/bfruntime_ext.cpp:1]: no copyright headers

File: backends/dpdk/control-plane/bfruntime_ext.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        addOneOf(dataJson, choicesDataJson, true /* mandatory */, false /* read-only */);
+    } else {
+        BUG("Invalid table type '%1%'", tableType);
+    }
+}
+
+void
+BFRuntimeSchemaGenerator::addActionSelectorGetMemberCommon(Util::JsonArray* tablesJson,
+                                        const ActionSelector& actionSelector) const {
+        auto* tableJson = initTableJson(actionSelector.get_mem_name,
```
Comment (mbudiu-vmw) [inline backends/dpdk/control-plane/bfruntime_ext.cpp:70]: inconsistent indentation

File: backends/dpdk/control-plane/bfruntime_ext.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        BF_RT_DATA_SELECTOR_GROUP_ID,
+        BF_RT_DATA_ACTION_MEMBER_STATUS,
+        BF_RT_DATA_MAX_GROUP_SIZE,
+        BF_RT_DATA_HASH_VALUE,
+    };
+    // Externs only for DPDK backend
+    struct ActionSelector;
+
+    void addDPDKExterns(Util::JsonArray* tablesJson,
+                                    Util::JsonArray* learnFiltersJson) const;
```
Comment (mbudiu-vmw) [inline backends/dpdk/control-plane/bfruntime_ext.h:58]: indentation seems inconsistent

File: control-plane/p4RuntimeArchStandard.h [hunk 5]
```diff
@@ -23,13 +23,201 @@ limitations under the License.
...
+using ::P4::ControlPlaneAPI::Helpers::setPreamble;
+
 namespace P4 {
 
 /** \addtogroup control_plane
  *  @{
  */
 namespace ControlPlaneAPI {
 
+namespace Standard {
```
Comment (mbudiu-vmw) [inline control-plane/p4RuntimeArchStandard.h:38]: I am assuming this has been moved from the cpp file. Github diff makes it difficult to check.

File: control-plane/bfruntime.h [hunk 6]
```diff
@@ -0,0 +1,497 @@
+#ifndef CONTROL_PLANE_BFRUNTIME_H_
```
Comment (mbudiu-vmw) [inline control-plane/bfruntime.h:1]: copyright

File: control-plane/bfruntime.h [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+        auto *tblObj = t->to<Util::JsonObject>();
+        auto tName = tblObj->get("name")->to<Util::JsonValue>()->getString();
+        if (tName == tblName) {
+            return tblObj;
+        }
+    }
+    return nullptr;
+}
+
+// See https://stackoverflow.com/a/33799784/4538702
```
Comment (mbudiu-vmw) [inline control-plane/bfruntime.h:57]: perhaps this should be in a library, such as json.h?

Comment (mbudiu-vmw) [reply control-plane/bfruntime.h:57]: BTW: there is already an escapeJson in cstring, but it's not exactly the same. Should that one be replaced?

File: control-plane/bfruntime.h [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+
+static const p4configv1::DirectMeter* findDirectMeter(const p4configv1::P4Info& p4info,
+                                                        P4Id meterId) {
+    const auto& meters = p4info.direct_meters();
+    return findP4InfoObject(meters.begin(), meters.end(), meterId);
+}
+
+}  // namespace Standard
+
+static Util::JsonObject* makeTypeInt(cstring type) {
```
Comment (mbudiu-vmw) [inline control-plane/bfruntime.h:152]: Why is this "Int"?

File: control-plane/bfruntime.h [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+        Util::JsonArray* annotations;
+
+        // static boost::optional<Register>
+        // from(const p4configv1::ExternInstance& externInstance);
+    };
+
+    void addMatchTables(Util::JsonArray* tablesJson) const;
+    virtual void addActionProfs(Util::JsonArray* tablesJson) const;
+    virtual bool addActionProfIds(const p4configv1::Table& table,
+                            Util::JsonObject* tableJson) const;
```
Comment (mbudiu-vmw) [inline control-plane/bfruntime.h:415]: in many places the continuation indentation is inconsistent

**Review by mbudiu-vmw — COMMENTED** [review]
I am really not the right person to review this, since I am not familiar with the control-plane part of p4c.
I made some small comments, but I will leave full approval to someone with better knowledge.

**Review by hanw — APPROVED** [review]
Looks good to me overall. @antoninbas if you have suggestion regarding the overall design, feel free to add more comments.

**Comment by amreshk** [comment]
@mbudiu-vmw , thanks for the review. Updated with fixes,
- Add missing copyright
- Fix indentation
- Change makeTypeInt to generic makeType
- Update `escapeJson()` in `cstring` library

Some of the dpdk test fails have changed fail signatures, there are no new failures. Will try to address them in a follow up PR.

File: control-plane/p4RuntimeArchStandard.h (hunk 5 above)
Comment (amreshk) [reply control-plane/p4RuntimeArchStandard.h:38]: Yes, this code needed to be in the header file to allow extending the class.

**Pull request merged**